File: Source/Extensions/Tessa.Extensions.Default.Client/Files/CycleGrouping.cs
using System;
using System.Linq;
using Tessa.Extensions.Default.Shared;
using Tessa.Files;
using Tessa.Localization;
using Tessa.Platform;
using Tessa.Platform.Formatting;
using Tessa.UI.Files;

namespace Tessa.Extensions.Default.Client.Files
{
    /// <summary>
    /// Группировка по циклу согласования
    /// </summary>
    /// <remarks>
    /// Класс-наследник может переопределить поведение класса, например,
    /// установить сортировку по умолчанию для данной группировки.
    /// </remarks>
    public class CycleGrouping : FileGrouping
    {
        #region Constructors

        public CycleGrouping(string name, string caption, bool isCollapsed = false)
            : base(name, caption, isCollapsed, new FileModelPropertyListener(nameof(IFile.Category), nameof(IFile.Origin)))
        {
        }

        #endregion

        #region Private Methods

        private static readonly FileDelegate<FileCaptionFunc> captionDelegate =
            new FileDelegate<FileCaptionFunc>(GetCaption, new[] { nameof(IFileObject.Name), nameof(IFile.Category), nameof(IFile.Origin) });

        private static string GetCaption(IFileViewModel viewModel)
        {
            IFile file = viewModel.Model;
            IFile origin = file.Origin;
            var category = file.Category;

            if (origin == null &&
                category != null &&
                viewModel.Collection
                    .Any(p => p.Model.Category != null &&
                              p.Model.Category.Equals(category)))
            {
                return
                    LocalizationManager.Format(
                        "$UI_Controls_FilesControl_Category", 
                        file.Name,
                        category.Caption);
            }

            if (origin == null &&
                file.Info.TryGetValue(CycleGroupingInfoKeys.CycleIDKey, out object cycleObj) &&
                cycleObj != null &&
                file.Info.TryGetValue(CycleGroupingInfoKeys.CreatedByNameKey, out object createtByNameObj) &&
                createtByNameObj != null &&
                file.Info.TryGetValue(CycleGroupingInfoKeys.CreatedKey, out object createdObj) &&
                createdObj != null)
            {
                return 
                    LocalizationManager.Format(
                        "$UI_Controls_FilesControl_Version",
                        file.Name,
                        file.Versions.Last.Number, 
                        (string)createtByNameObj);
            }

            if (origin != null)
            {
                return string.Format(
                    LocalizationManager.GetString("UI_Controls_FilesControl_ByCopy_Format"),
                    file.Name,
                    LocalizationManager.GetString("UI_Controls_FilesControl_ByCopy_Copy"),
                    file.Versions.Last.CreatedByName,
                    FormattingHelper.FormatDateTime(file.Versions.Last.Created));
            }

            return file.Name;
        }

        #endregion

        #region Constants

        /// <summary>
        /// Группа "Согласуемые документы".
        /// </summary>
        public const string DocumentsOnApprovalGroupName = "DocumentsOnApprovalGroup";

        public const string CycleGroupName = "CycleGroup";

        #endregion

        #region Base Overrides

        /// <doc path='info[@type="IFileGrouping" and @item="GetGroupInfo"]'/>
        protected override FileGroupInfo GetGroupInfoCore(IFileViewModel viewModel)
        {
            if (viewModel.Model.Info.TryGetValue(CycleGroupingInfoKeys.CycleIDKey, out object cycleNumberObj) &&
                cycleNumberObj != null &&
                viewModel.Model.Info.TryGetValue(CycleGroupingInfoKeys.CycleOrderKey, out object cycleOrderObj) &&
                cycleOrderObj != null &&
                viewModel.Model.Info.TryGetValue(CycleGroupingInfoKeys.MaxCycleNumberKey, out object maxCycleNumberObj) &&
                maxCycleNumberObj != null)
            {
                int length = Math.Max(1, maxCycleNumberObj.ToString()?.Length ?? 1);
                string cycleNumberString = cycleNumberObj.ToString();
                
                return new 
                    FileGroupInfo(
                        CycleGroupName + cycleNumberString,
                        LocalizationManager.Format("$UI_Controls_FilesControl_CycleGroup", cycleNumberString), 
                        ((int)cycleOrderObj + 1).ToString("D" + length));
            }

            return new FileGroupInfo(DocumentsOnApprovalGroupName, LocalizationManager.GetString("UI_Controls_FilesControl_DocumentsOnApprovalGroup"), "0");
        }

        
        /// <doc path='info[@type="FileControlObject" and @item="Attach"]'/>
        protected override void Attach(IFileViewModel viewModel)
        {
            base.Attach(viewModel);

            viewModel.CaptionDelegateManager.Set(captionDelegate);
        }


        /// <doc path='info[@type="FileControlObject" and @item="Detach"]'/>
        protected override void Detach(IFileViewModel viewModel)
        {
            FileDelegate<FileCaptionFunc> restoredDelegate = viewModel.CaptionDelegateManager.Restore();
            if (restoredDelegate != captionDelegate)
            {
                throw new InvalidOperationException($"{nameof(IFileViewModel)}.{nameof(IFileViewModel.CaptionDelegateManager)} stack is damaged.");
            }

            base.Detach(viewModel);
        }
        
        #endregion
    }
}

File: WebClient SDK/src/default/files/cycleGrouping.ts
import { FileGrouping, FileViewModel, GroupInfo } from 'tessa/ui/cards/controls';
import { tryGetFromInfo } from 'tessa/ui';
import { LocalizationManager } from 'tessa/localization';
import { formatToString, formatDate } from 'tessa/platform/formatting';
import { FileCategory } from 'tessa/files';

// tslint:disable: triple-equals

export class CycleGrouping extends FileGrouping {
  //#region ctor

  constructor(name: string, caption: string, isCollapsed: boolean = false) {
    super(name, caption, isCollapsed);
  }

  //#endregion

  //#region methods

  public getGroupInfo(file: FileViewModel): GroupInfo {
    const cycleNumber = tryGetFromInfo<number>(file.model.info, 'KrCycleID');
    const cycleOrder = tryGetFromInfo<number>(file.model.info, 'KrCycleorder');
    const maxCyclewNumber = tryGetFromInfo<number>(file.model.info, 'KrMaxCycleNumber');
    if (cycleNumber != undefined && cycleOrder != undefined && maxCyclewNumber != undefined) {
      return {
        groupId: `CycleGroup${cycleNumber}`,
        groupCaption: LocalizationManager.instance.format(
          '$UI_Controls_FilesControl_CycleGroup',
          cycleNumber
        ),
        order: cycleOrder + 1
      };
    }

    return {
      groupId: 'DocumentsOnApprovalGroup',
      groupCaption: LocalizationManager.instance.localize(
        '$UI_Controls_FilesControl_DocumentsOnApprovalGroup'
      ),
      order: 0
    };
  }

  public clone(): CycleGrouping {
    return new CycleGrouping(this.name, this.caption, this.isCollapsed);
  }

  public attach(file: FileViewModel) {
    super.attach(file);
    file.captionDelegate.set(CycleGrouping.getCaption);
  }

  public detach(file: FileViewModel) {
    super.detach(file);
    const restoredDelegate = file.captionDelegate.restore();
    if (restoredDelegate !== CycleGrouping.getCaption) {
      throw new Error('FileViewModel.captionDelegate stack is damaged.');
    }
  }

  private static getCaption(viewModel: FileViewModel): string {
    const file = viewModel.model;
    const origin = file.origin;
    const category = file.category;

    if (
      !origin &&
      category &&
      viewModel.collection.some(
        x => x.model.category && FileCategory.equals(x.model.category, category)
      )
    ) {
      return LocalizationManager.instance.format(
        '$UI_Controls_FilesControl_Category',
        file.name,
        category.caption
      );
    }

    const cycle = tryGetFromInfo<number>(file.info, 'KrCycleID');
    const createdByName = tryGetFromInfo<string>(file.info, 'KrCreatedByName');
    const created = tryGetFromInfo<string>(file.info, 'KrCreated');

    if (!origin && cycle != undefined && createdByName != undefined && created != undefined) {
      return LocalizationManager.instance.format(
        '$UI_Controls_FilesControl_Version',
        file.name,
        file.lastVersion.number,
        createdByName
      );
    }

    if (origin) {
      const format = LocalizationManager.instance.localize(
        '$UI_Controls_FilesControl_ByCopy_Format'
      );
      const copy = LocalizationManager.instance.localize('$UI_Controls_FilesControl_ByCopy_Copy');
      return formatToString(
        format,
        file.name,
        copy,
        file.lastVersion.createdByName,
        formatDate(file.lastVersion.created)
      );
    }

    return file.name;
  }

  //#endregion
}
